shared/tinyusb: Add macro to override TinyUSB callbacks.

iabdalkader opened the pull request:
### Summary

I need to redirect some of the TinyUSB callbacks, but MicroPython implements them exclusively. So far I've been using linker `--wrap` with great success, sadly TinyUSB 0.19.0 adds weak symbols for callbacks which breaks the `--wrap` trick.
This patch adds `MICROPY_TINYUSB_CALLBACK` macro that by default expands to the callback name unchanged. Users can define this macro to add a prefix (e.g., mp_) to callback implementations, to redirect or completely override MicroPython's TinyUSB callbacks.

### Other changes

Add hook to mp_usbd_task. This adds an empty hook that gets called when running the USB task.

### Testing

`MICROPY_TINYUSB_CALLBACK` macro expands to the callback name unchanged, I've double checked the map file and in both cases (default, or defined to `mp_`) the symbols have the right names. Let me know if more testing needed.

### Trade-offs and Alternatives

This is the most direct way I could think of, however:

* `--wrap` no longer works with TinyUSB 0.19.0.
* `objcopy` to rename symbols. I had a hard time trying to get this to work, it requires complex rules somewhere, but might be a viable option if this is closed.
* Hooks?
```C
  // In mp_usbd_cdc.c
  void tud_cdc_rx_cb(uint8_t itf) {
      // Call user hook if registered
      if (mp_usbd_cdc_rx_hook != NULL) {
          if (mp_usbd_cdc_rx_hook(itf)) {
              return; // Hook handled it
          }
      }
      // Default implementation
      mp_usbd_cdc_rx_cb_default(itf);
  }
```
* `MICROPY_EXCLUDE_SHARED_TINYUSB_USBD_CDC` seems to have been needed at some point, but it requires a full implementation of those callbacks. I only need to redirect/override some of them.

* Make mp callbacks weak?
1- Rename those callbacks to `mp_tud_cdc_x`
2- Implement `tud_cdc_`  as yet another weak symbol that calls `mp_tud_cdc_x`
3- Now we can provide strong symbols that override them, and call `mp_tud_cdc_x` if needed.

This depends on the link order, I think the shared MP objects always come first though.

dpgeorge commented on shared/tinyusb/mp_usbd_cdc.c:
```diff
@@ -68,7 +72,7 @@ uintptr_t mp_usbd_cdc_poll_interfaces(uintptr_t poll_flags) {
 }
 
-void tud_cdc_rx_cb(uint8_t itf) {
+void MICROPY_TINYUSB_CALLBACK(tud_cdc_rx_cb)(uint8_t itf) {
```
Note that `tud_cdc_rx_cb` is called from `mp_usbd_cdc_poll_interfaces()`... not sure if you want the wrapped version to be called, or the raw one?

iabdalkader replied:
Good catch! The wrapped version should be called. However, I also wrap:
```C
           -Wl,--wrap=mp_hal_stdio_poll \
           -Wl,--wrap=mp_hal_stdout_tx_strn
```
So I don't think it's a problem, but I will fix it anyway.

iabdalkader committed:
shared/tinyusb: Add macro to override TinyUSB callbacks.

Add wrapper macros that by default expand to the callback name.
Users can define this macro to add a prefix (e.g., mp_) to
callback implementations, to redirect or completely override
MicroPython's TinyUSB callbacks.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

dpgeorge reviewed (approved):
Looks good now!
